### PR TITLE
Space policy UI

### PR DESCRIPTION
**Pull request by jreidinger**
## Problem

There is already support in backend for space policy for storage proposal, but UI is missing.

- https://trello.com/c/VjNLDTd6/3410-5-storage-ui-to-select-the-settings-to-make-space-except-custom

## Solution

Implement UI.

What is missing:

- [x] Good representation of the affected devices at the form (including some code refactoring at `device-utils.jsx`)
- [x] Better font layout and style
- [x] Improve wording
- [x] Support for singular / plural (number of affected devices)

## Testing

What is missing:

- [ ] Manual tests of a full installation (maybe better done after merging, together with other stuff currently at integration phase)
- [x] Unit tests

## Screenshots

space

police

### CODE REVIEW

**Comment by dgdavid** [comment]
Although PR is in draft mode yet, I wanted to drop here a thought: I wouldn't use a popup for displaying such a "selector". Instead, I'd put it directly in the page. I mean, if the selector is just the group of "buttons" which only one selected at the time. 

We can talk about it tomorrow morning.

File: web/src/components/storage/ProposalSettingsSection.jsx [hunk 1]
```diff
@@ -509,6 +510,162 @@ const EncryptionPasswordField = ({
...
+    onSubmitProp(spacePolicy);
+  };
+
+  return (
+    <Form id={id} onSubmit={onSubmit}>
+      <div className="split">
+        <span>{_("Policy to make enough free space")}</span>
+        <ToggleGroup isCompact>
+          <ToggleGroupItem
+            text={_("Delete Anything")}
```
Comment (mvidner) [inline web/src/components/storage/ProposalSettingsSection.jsx:555]:
```suggestion
            text={_("Delete Everything")}
```

Help text?

> All contents (*) will be deleted and lost.
> Make sure that the correct target disk is selected, that you have copied any data that you might still need from the target disk(s).

For (*), if this applies to some disks only, insert
> of the target disk `/dev/sda`

File: web/src/components/storage/ProposalSettingsSection.jsx [hunk 2]
```diff
@@ -509,6 +510,162 @@ const EncryptionPasswordField = ({
...
+            onClick={selectDelete}
+          />
+          <ToggleGroupItem
+            text={_("Resize Only")}
+            buttonId="resizePolicy"
+            isSelected={spacePolicy === "resize"}
+            onClick={selectResize}
+          />
+          <ToggleGroupItem
+            text={_("Keep Existing")}
```
Comment (mvidner) [inline web/src/components/storage/ProposalSettingsSection.jsx:567]:
```suggestion
            text={_("Keep Everything")}
```

Help text

> Existing partitions and their data* will not be touched. There needs to be enough unpartitioned space on the target disk already.
>
> *: If there is an EFI boot partition, its contents will be modified to allow booting the system being installed.

am I correct in this?

File: web/src/components/storage/ProposalSettingsSection.jsx [hunk 3]
```diff
@@ -509,6 +510,162 @@ const EncryptionPasswordField = ({
...
+        <span>{_("Policy to make enough free space")}</span>
+        <ToggleGroup isCompact>
+          <ToggleGroupItem
+            text={_("Delete Anything")}
+            buttonId="deletePolicy"
+            isSelected={spacePolicy === "delete"}
+            onClick={selectDelete}
+          />
+          <ToggleGroupItem
+            text={_("Resize Only")}
```
Comment (mvidner) [inline web/src/components/storage/ProposalSettingsSection.jsx:561]:
```suggestion
            text={_("Shrink Only")}
```

Help text

> Some of the existing partitions will be resized to make them smaller, keeping their contents, just enough to make space for the system being installed.

I wonder, does this really work automatically? Last time I used something like this, I wanted to specify **how much** space would be allocated to Linux and how much the fresh and mostly empty Windows system would get to keep.

File: web/src/components/storage/ProposalSettingsSection.jsx [hunk 4]
```diff
@@ -509,6 +510,162 @@ const EncryptionPasswordField = ({
...
+  const onSubmitForm = (policy) => {
+    onChange(policy);
+    setSpacePolicy(policy);
+    closeForm();
+  };
+
+  const SpacePolicySettingsButton = () => {
+    return (
+      <Tooltip
+        content={_("Configure the Space Policy")}
```
Comment (mvidner) [inline web/src/components/storage/ProposalSettingsSection.jsx:611]: Go! For!! Mars!!! (SCNR)

Maybe use _Storage Space (Policy)_ instead of just _Space (Policy)_? Everywhere?

File: web/src/components/storage/ProposalSettingsSection.jsx [hunk 5]
```diff
@@ -509,6 +510,162 @@ const EncryptionPasswordField = ({
...
+
+  if (isLoading) return <Skeleton width="25%" />;
+
+  let text = "";
+  switch (spacePolicy) {
+    case "delete":
+      text = _("Delete everything");
+      break;
+    case "resize":
+      text = _("Resize as needed");
```
Comment (mvidner) [inline web/src/components/storage/ProposalSettingsSection.jsx:631]:
```suggestion
      text = _("Shrink only");
```

File: web/src/components/storage/ProposalSettingsSection.jsx [hunk 6]
```diff
@@ -509,6 +510,162 @@ const EncryptionPasswordField = ({
...
+      break;
+    case "keep":
+      text = _("Keep existing stuff");
```
Comment (mvidner) [inline web/src/components/storage/ProposalSettingsSection.jsx:634]:
```suggestion
      text = _("Keep everything");
```

Here I am changing the labels to match the popup (except for capitalization, not sure about that)

File: web/src/components/storage/ProposalSettingsSection.jsx [hunk 7]
```diff
@@ -509,6 +510,162 @@ const EncryptionPasswordField = ({
...
+        position="right"
+      >
+        <button aria-label={_("Space Policy settings")} className="plain-control" onClick={openForm}>
+          <Icon name="tune" size={24} />
+        </button>
+      </Tooltip>
+    );
+  };
+
+  if (isLoading) return <Skeleton width="25%" />;
```
Comment (mvidner) [inline web/src/components/storage/ProposalSettingsSection.jsx:623]: This looks like it could be moved to the top of `SpacePolicyField`, an early return

File: web/src/components/storage/ProposalSettingsSection.jsx [hunk 8]
```diff
@@ -509,6 +510,162 @@ const EncryptionPasswordField = ({
...
+      {value_text}
+      <SpacePolicySettingsButton />
+      <Popup
+        aria-label={_("Space Policy settings")}
+        title={_("Space Policy")}
+        isOpen={isFormOpen}
+      >
+        <SpacePolicySettingsForm
+          id="spacePolicySettingsForm"
+          settings={spacePolicy}
```
Comment (mvidner) [inline web/src/components/storage/ProposalSettingsSection.jsx:652]: Here I find confusing that `settings` is not an object but a single setting, a scalar.
I would expect `policy={spacePolicy}` or `settings={ {spacePolicy} }` (if I'm getting the single-valued object syntax right)

File: web/src/components/storage/ProposalSettingsSection.jsx [hunk 9]
```diff
@@ -509,6 +510,100 @@ const EncryptionPasswordField = ({
...
+
+  const onSubmitForm = (policy) => {
+    onChange(policy);
+    setSpacePolicy(policy);
+    closeForm();
+  };
+
+  if (isLoading) return <Skeleton width="25%" />;
+
+  const text = sprintf(_("Find space at the %d chosen disks"), settings.installationDevices.length);
```
Comment (mvidner) [inline web/src/components/storage/ProposalSettingsSection.jsx:577]:
```suggestion
  const text = sprintf(_("Find space in the %d chosen disks"), settings.installationDevices.length);
```

Comment (mvidner) [reply web/src/components/storage/ProposalSettingsSection.jsx:577]: with plural forms:

```suggestion
  const text = sprintf(
    n_("Find space in the %d chosen disk", "Find space in the %d chosen disks", settings.installationDevices.length),
    settings.installationDevices.length
  );
```

File: web/src/components/storage/space-policy-utils.jsx [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+const SpacePolicyButton = ({ policy, onClick = noop }) => {
+  const text = (policy) => {
+    switch (policy) {
+      case "delete":
+        return _("deleting its current content");
+      case "resize":
+        return _("shrinking existing partitions");
+      case "keep":
+        return _("without modifying any partition");
```
Comment (mvidner) [inline web/src/components/storage/space-policy-utils.jsx:116]: These 3 are sentence _fragments_. Translators need to know the rest of the sentence, otherwise grammar mismatches are likely.

**Comment by dgdavid** [comment]
> Because I thought you're going to complaint that sight people can miss for a while the _important_

BTW, just recording here for checking it tomorrow, another option in both cases would be to force the browser not hide the scrollbar in the Hint area. I think it helps a lot to quickly spot that there is hidden content (and a rough clue about how many). But not sure at this moment how doable it is... Something for tomorrow.

**Comment by dgdavid** [comment]
JFTR, this morning @ancorgs and me discussed about a new idea using the PF/Hint plus a disclosure component for letting the user display disks only if interested. Less invasive, we think.

File: web/src/components/storage/space-policy-utils.jsx [hunk 11]
```diff
@@ -105,21 +108,64 @@ const SpacePolicySelector = ({ value, onChange = noop }) => {
...
         return _("without modifying any partition");
     }
     console.log("Unsupported value " + policy);
     return "error";
   };
 
-  return <Button variant="link" isInline onClick={onClick}>{text(policy)}</Button>;
+  const num = devices.length;
+
+  return <Button variant="link" isInline onClick={onClick}>{text(policy, num)}</Button>;
```
Comment (dgdavid) [inline web/src/components/storage/space-policy-utils.jsx:136]: No strong opinion to be honest, but when declaring and using functions like `text` here in the middle of the JSX I tend to use an internal React component instead. I.e., 

```js
const Text = ({ policy, num }) => { .... }
....
<Button ....>
  <Text policy={policy} num={devices.length} />
</Button>
```

What is more, you could even go further and remove these props since it's an internal component.

But it's just another way to do it, to ending up having a JSX block with less interpolation.

File: web/src/client/storage.js [hunk 12]
```diff
@@ -320,13 +323,28 @@ class ProposalManager {
...
+
       return {
         settings: {
           bootDevice: proxy.BootDevice,
           lvm: proxy.LVM,
+          spacePolicy: proxy.SpacePolicy,
           systemVGDevices: proxy.SystemVGDevices,
           encryptionPassword: proxy.EncryptionPassword,
           volumes: proxy.Volumes.map(this.buildVolume),
+          installationDevices: buildInstallationDevices(proxy, systemDevices)
```
Comment (joseivanlopez) [inline web/src/client/storage.js:347]: Installation devices is not a setting but some information we can get from the settings. I would extract it from settings, offerting a separate funtion to get the list: `client.installationDevices()`.

File: web/src/components/storage/space-policy-utils.jsx [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+
+/**
+ * Content for a space policy item
+ * @component
+ *
+ * @param {Object} props
+ * @param {Locale} props.locale
+ */
+const PolicyItem = ({ policy }) => {
+  const Title = () => {
```
Comment (joseivanlopez) [inline web/src/components/storage/space-policy-utils.jsx:54]: NP: `Title` and `Description` could be helper functions instead of components returning a string.

File: web/src/components/storage/device-utils.jsx [hunk 14]
```diff
@@ -229,11 +229,11 @@ const DeviceItem = ({ device }) => {
  * @param {Object} props
  * @param {StorageDevice[]} props.devices - Devices to show.
  */
-const DeviceList = ({ devices }) => {
+const DeviceList = ({ devices, ...props }) => {
```
Comment (joseivanlopez) [inline web/src/components/storage/device-utils.jsx:232]: It is a bit weird to pass list props to each item. Anyway, we need to rethink selectors and also these lists. I would add a note.

File: web/src/client/storage.js [hunk 15]
```diff
@@ -320,13 +323,28 @@ class ProposalManager {
...
+        const findDevice = (devices, name) => {
+          const device = devices.find(d => d.name === name);
+
+          if (device === undefined) console.log("D-Bus object not found: ", name);
+
+          return device;
+        };
+
+        const names = proxy.SystemVGDevices.filter(n => n !== proxy.BootDevice).concat([proxy.BootDevice]);
+        return names.map(dev => findDevice(devices, dev));
```
Comment (joseivanlopez) [inline web/src/client/storage.js:336]: I would remove undefined values from the list.

File: web/src/components/storage/space-policy-utils.jsx (hunk 13 above)
Comment (dgdavid) [reply web/src/components/storage/space-policy-utils.jsx:54]: NP: I told Áncor just the contrary :)

**Comment by dgdavid** [comment]
@ancorgs, @mvidner, @joseivanlopez,

I've sent a few commits. I think the space policy dialog looks better now, let me know if there is something you believe should be changed for accepting this PR from the UI perspective.

Remember that in general selectors need more love, but it deserves a separated PR.

BTW, I did a last change but didn't send it because it's something I'd do once #881 and this PR are merged. It's about making selectors more compact by dropping the space between their options. If you are curious, see below the diff and a few screenshots


<details><summary>Click to show/hide the screenshots</summary>

---

Screenshot from 2023-11-29 20-22-00
Screenshot from 2023-11-29 20-21-42
Screenshot from 2023-11-29 20-21-37
</details>


<details><summary>Click to show/hide the (NOT SENT) diff</summary>

---

```diff
diff --git a/web/src/assets/styles/blocks.scss b/web/src/assets/styles/blocks.scss
index 30ac1f73..7218d7ce 100644
--- a/web/src/assets/styles/blocks.scss
+++ b/web/src/assets/styles/blocks.scss
@@ -267,11 +267,21 @@ span.notification-mark[data-variant="sidebar"] {
 ul[data-type="agama/list"] {
   li {
     border: 2px solid var(--color-gray-dark);
-    border-radius: 5px;
     padding: var(--spacer-normal);
     text-align: start;
     background: var(--color-gray-light);
-    margin-block-end: var(--spacer-small);
+
+    &:first-child {
+      border-radius: 5px 5px 0 0;
+    }
+
+    &:last-child {
+      border-radius: 0 0 5px 5px;
+    }
+
+    &:nth-child(n+2) {
+      border-top: 0;
+    }
   }
```
</details>

File: web/src/client/storage.js (hunk 12 above)
Comment (joseivanlopez) [reply web/src/client/storage.js:347]: Added a comment.

File: web/package/cockpit-agama.changes [hunk 16]
```diff
@@ -1,3 +1,9 @@
+-------------------------------------------------------------------
+Thu Nov 30 15:19:38 UTC 2023 - José Iván López González <jlopez@suse.com>
+
+- Allow selecting the storage policy to make free space
```
Comment (jreidinger) [inline web/package/cockpit-agama.changes:4]: maybe I would rephrase it as `Allow selecting the storage policy to make space for installation`

File: web/src/client/storage.js [hunk 17]
```diff
@@ -320,13 +324,32 @@ class ProposalManager {
...
+        const findDevice = (devices, name) => {
+          const device = devices.find(d => d.name === name);
+
+          if (device === undefined) console.log("D-Bus object not found: ", name);
+
+          return device;
+        };
+
+        const names = proxy.SystemVGDevices.filter(n => n !== proxy.BootDevice).concat([proxy.BootDevice]);
+        return names.map(dev => findDevice(devices, dev)).filter(Boolean);
```
[the remark this answers is not in the document]
Comment (jreidinger) [reply web/src/client/storage.js:337]: also maybe just using unique method can help a bit with readability ( aka `const unique_names = [...new Set(names)]`

File: web/src/components/overview/StorageSection.jsx [hunk 18]
```diff
@@ -29,19 +29,39 @@ import { deviceLabel } from "~/components/storage/utils";
...
+        // example: "/dev/sda, 20 GiB"
+        return _("Install using device %s shrinking existing partitions as needed");
+      case "keep":
+        // TRANSLATORS: %s will be replaced by the device name and its size,
+        // example: "/dev/sda, 20 GiB"
+        return _("Install using device %s without modifying existing partitions");
+    }
+    // TRANSLATORS: %s will be replaced by the device name and its size,
+    // example: "/dev/sda, 20 GiB"
+    return _("Install using device %s and deleting all its content");
```
Comment (jreidinger) [inline web/src/components/overview/StorageSection.jsx:60]: to be honest I am not fan of those implivit fallthrough for enum values. I would add case also for `"delete"` so if we support custom in future it can properly write to console something like "unknown value" and you future Ivan will thank you.

Comment (ancorgs) [reply web/src/components/overview/StorageSection.jsx:60]: Sure. But take into account the full `ProposalSummary` component is meant to be rewritten in the short term.

File: web/src/components/storage/ProposalSettingsSection.jsx [hunk 19]
```diff
@@ -509,6 +513,103 @@ const EncryptionPasswordField = ({
...
+            type="submit"
+          >
+            {_("Accept")}
+          </Popup.Confirm>
+          <Popup.Cancel onClick={closeForm} />
+        </Popup.Actions>
+      </Popup>
+    </div>
+  );
+};
```
[the remark this answers is not in the document]
Comment (joseivanlopez) [reply web/src/components/storage/ProposalSettingsSection.jsx:611]: I would evaluate it later if this section continues growing. Right now all its `*Field` components are defined there.

**Review by dgdavid — APPROVED** [review]
